clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
moreMemory.await() is a blocking call. So this means that suppose we had condition variables C1,C2,C3,C4 in the queue and if C4 has timed out, it means that C1 has to be timed out. So when C1 times out, the if condition that checks for waitingTimeElaspsed will remove that condition variable if we change the remove to this.waiters.remove(moreMemory) instead of just doing this.waiters.removeLast() as in this PR. I think that should solve the issue that you pointed out.